Sources/Bot/Services/DispatchService.swift

```diff
@@ -0,0 +1,198 @@
+                case let .status(event):
+                    self?.statusChecksDidChange(event: event)
+                case .ping:
+                    break
+                }
+            }
+    }
+
+    private func dispatchInitial(pullRequests: [PullRequest]) {
+        let dispatchTable = Dictionary(grouping: pullRequests) { pullRequest in pullRequest.target.ref }
+        for (branch, pullRequestsForBranch) in dispatchTable {
+            makeMergeService(targetBranch: branch, scheduler: self.scheduler, initialPullRequests: pullRequestsForBranch)
+        }
+    }
+
+    private func pullRequestDidChange(event: PullRequestEvent) {
+        logger.log("📣 Pull Request did change \(event.pullRequestMetadata) with action `\(event.action)`")
+        let targetBranch = event.pullRequestMetadata.reference.target.ref
+        if let service = mergeServices[targetBranch] {
+            service.pullRequestChangesObserver.send(value: (event.pullRequestMetadata, event.action))
+        } else {
+            makeMergeService(targetBranch: targetBranch, scheduler: self.scheduler, initialPullRequests: [event.pullRequestMetadata.reference])
+        }
+    }
+
+    private func statusChecksDidChange(event: StatusEvent) {
+        // No way to know which MergeService this event is supposed to be for – isRelative(toBranch:) only checks for head branch not target so not useful here
+        // So we're sending it to all MergeServices, and they'll filter them themselves based on their own queues
+        for mergeServiceForBranch in mergeServices.values {
+            mergeServiceForBranch.statusChecksCompletionObserver.send(value: event)
+        }
+    }
+
+    @discardableResult
+    private func makeMergeService(targetBranch: String, scheduler: DateScheduler, initialPullRequests: [PullRequest] = []) -> MergeService {
+        let mergeService = MergeService(
+            targetBranch: targetBranch,
+            integrationLabel: integrationLabel,
+            topPriorityLabels: topPriorityLabels,
+            requiresAllStatusChecks: requiresAllStatusChecks,
+            statusChecksTimeout: statusChecksTimeout,
+            initialPullRequests: initialPullRequests,
+            logger: logger,
+            gitHubAPI: gitHubAPI,
+            scheduler: scheduler
+        )
+        self.mergeServices[targetBranch] = mergeService
+        self.healthcheck.startMonitoring(mergeServiceHealthcheck: mergeService.healthcheck)
+        mergeServiceLifecycleObserver?.send(value: .created(mergeService))
+        mergeService.state.producer
+            .observe(on: scheduler)
+            .startWithValues { [weak self, service = mergeService] state in
+                self?.mergeServiceLifecycleObserver?.send(value: .stateChanged(service))
+                if state.status == .idle {
+                    self?.mergeServices[targetBranch] = nil
+                    self?.mergeServiceLifecycleObserver?.send(value: .destroyed(service))
```

Comment:
another idea is to call these `.send` methods in the setter of `mergeServices`? it might be a bit tricky though to find what key actually changed...

Reply:
Yeah I thought about that but it felt more complicated that it was worth to have to do the diff manually and detect new keys instead of directly trigger them at the point we delete them ourselves – `mergeServices` is private after all so nothing other that this code should alter it.

I can explore the solution about observing `mergeServices` changes but that would require us to transform it to a `Property` instead of a single dictionary, just to be able to observe it and `combinePrevious` and do some math, and I'm seeing more risks of doing that wrong than benefits tbh, no?

Reply:
yes, it seems more complicated this way